File: 662_Maximum_Width_of_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int dfs(TreeNode* root, int level, int order, vector<pair<int, int>> &record) {
        if (root == NULL) return 0;
        if (record.size() == level) record.push_back({order, order});
        else record[level].second = order;
        return max({record[level].second - record[level].first + 1,
                    dfs(root->left, level + 1, order * 2, record),
                    dfs(root->right, level + 1, order * 2 + 1, record)});
    }
    int widthOfBinaryTree(TreeNode* root) {
        return dfs(root, 0, 1, vector<pair<int, int>>() = {});
    }
};

File: 662_Maximum_Width_of_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def widthOfBinaryTree(self, root: TreeNode) -> int:
        width, level = 0, [(root, 0)]
        while len(level):
            width = max(width, level[-1][1] - level[0][1] + 1)
            next_level = []
            for node, idx in level:
                if node.left:
                    next_level.append((node.left, idx * 2))
                if node.right:
                    next_level.append((node.right, idx * 2 + 1))
            level = next_level
        return width
